compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -864,7 +868,13 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
       accu(Set.empty, selType)
     }
 
-    cases mapconserve (typedCase(_, pt, selType, gadtSyms))
+    val tpdCases = cases mapconserve (typedCase(_, pt, selType, gadtSyms))
+
+    val phantomBranches = tpdCases.count(_.body.tpe.isPhantom)
+    if (phantomBranches != 0 && phantomBranches != tpdCases.size)
+      ctx.error(MatchPhantom(), tpdCases.head.pos)
+
```

Comment:
Indeed. The `topClass` is the solution. 

The same applies to the `if/else` branches.